### ISSUE TITLE
Function Overloads Issue

### ISSUE THREAD

**Issue by audinue** [Microsoft/TypeScript#17508]
**TypeScript Version:**  2.4.0 / nightly (2.5.0-dev.201xxxxx)

Found something weird this night.

**Code**

```ts
interface Entity {
    id: string;
}

function add(entities: Entity[]): void;
function add(entity: Entity): void;
function add(entitiesOrEntity: Entity[] | Entity): void {
}

// Copy-pasted, renamed 'add' to 'reset'.
function reset(entities: Entity[]): void;
function reset(entity: Entity): void;
function reset(entitiesOrEntity: Entity[] | Entity): void {
    add(entitiesOrEntity); // Error?
}
```

**Expected behavior:**
Doesn't show errors because the signatures are compatible.

**Actual behavior:**
```
Argument of type 'Entity | Entity[]' is not assignable to parameter of type 'Entity'.
  Type 'Entity[]' is not assignable to type 'Entity'.
    Property 'id' is missing in type 'Entity[]'.
```

See also TypeScript playground [link](http://www.typescriptlang.org/play/index.html#src=interface%20Entity%20%7B%0D%0A%20%20%20%20id%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20add(entities%3A%20Entity%5B%5D)%3A%20void%3B%0D%0Afunction%20add(entity%3A%20Entity)%3A%20void%3B%0D%0Afunction%20add(entitiesOrEntity%3A%20Entity%5B%5D%20%7C%20Entity)%3A%20void%20%7B%0D%0A%7D%0D%0A%0D%0Afunction%20reset(entities%3A%20Entity%5B%5D)%3A%20void%3B%0D%0Afunction%20reset(entity%3A%20Entity)%3A%20void%3B%0D%0Afunction%20reset(entitiesOrEntity%3A%20Entity%5B%5D%20%7C%20Entity)%3A%20void%20%7B%0D%0A%20%20%20%20add(entitiesOrEntity)%3B%20%2F%2F%20Error%3F%0D%0A%7D).

**Comment by aluanhaddad** [comment]
Which signature should be chosen, there is not an available one that matches. Just use a union type. 
The overloads are not serving a purpose in this case.

**Comment by audinue** [comment]
Would you mind to explain what do you mean about "The overloads are not serving a purpose", please?

**Comment by NN---** [comment]
@audinue Just remove the overloads. You don't need them if you specify union type:

```ts
interface Entity {
    id: string;
}

function add(entitiesOrEntity: Entity[] | Entity): void {
}

function reset(entitiesOrEntity: Entity[] | Entity): void {
    add(entitiesOrEntity); // Error?
}
```

**Comment by audinue** [comment]
@NN--- Thank you for the explanation 😄
